flink-connectors/flink-hadoop-compatibility/src/main/java/org/apache/flink/hadoopcompatibility/mapreduce/wrapper/HadoopProxyMapper.java

```diff
@@ -0,0 +1,388 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.hadoopcompatibility.mapreduce.wrapper;
+
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.util.Collector;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.io.RawComparator;
+import org.apache.hadoop.mapreduce.Counter;
+import org.apache.hadoop.mapreduce.InputFormat;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.JobID;
+import org.apache.hadoop.mapreduce.Mapper;
+import org.apache.hadoop.mapreduce.OutputCommitter;
+import org.apache.hadoop.mapreduce.OutputFormat;
+import org.apache.hadoop.mapreduce.Partitioner;
+import org.apache.hadoop.mapreduce.Reducer;
+import org.apache.hadoop.mapreduce.TaskAttemptID;
+import org.apache.hadoop.security.Credentials;
+
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.net.URI;
+
+/**
+ * This is a proxy maps a Hadoop Mapper (mapreduce API) which use reflection to get and invoke some method.
+ */
+public class HadoopProxyMapper<KEYIN, VALUEIN, KEYOUT, VALUEOUT> extends Mapper<KEYIN, VALUEIN, KEYOUT, VALUEOUT> {
+
+	private final Mapper delegatedMapper;
+	private Method mapMethod;
+	private Method setupMethod;
+	private Method cleanupMethod;
+
+	public HadoopProxyMapper(Mapper delegatedMapper) {
+		this.delegatedMapper = Preconditions.checkNotNull(delegatedMapper);
+		try {
+			mapMethod = delegatedMapper.getClass().getDeclaredMethod("map", Object.class, Object.class, Mapper.Context.class);
+			mapMethod.setAccessible(true);
+		} catch (NoSuchMethodException e) {
+			mapMethod = null;
+		}
+
+		try {
+			setupMethod = delegatedMapper.getClass().getDeclaredMethod("setup", Context.class);
+			setupMethod.setAccessible(true);
+		} catch (NoSuchMethodException e) {
+			setupMethod = null;
+		}
+
+		try {
+			cleanupMethod = delegatedMapper.getClass().getDeclaredMethod("cleanup", Context.class);
+			cleanupMethod.setAccessible(true);
+		} catch (NoSuchMethodException e) {
+			cleanupMethod = null;
+		}
+	}
+
+	@Override
+	public void setup(Context context) throws IOException, InterruptedException {
+		super.setup(context);
+		if (setupMethod != null) {
+			try {
```

Comment:
If we move this to `HadoopMapFunction`, we can remove the `try` completely as `MapFunction.open()` throws `Exception`.